### PR TITLE
ztest: add test summary support.

### PR DESCRIPTION
**Pull request by smrtos**
This will be a series of commits.

**[Commit 1]**:  Add config switch for modes of test output
Support verbose or one-line summary at test suite level.
Support verbose or no output at test function level.

Totally 4 combinations configurable:
- function verbose + suite verbose
- function verbose + suite one-line
- no function output + suite verbose
- no function output + suite one-line

**[Commit 2]**:   Add test summary after all suites finish running

  Add test summary after all test suites finish running.
  The summary can be one-line or verbose, which is configured
  with CONFIG_ZTEST_VERBOSE_SUMMARY. The one-line summary covers
  overall suite stats. The verbose summary covers each test
  function within the suite besides the one-line summary.

Signed-off-by: Ming Shao <ming.shao@intel.com>

### CODE REVIEW

**Comment by yperess** [comment]
Can you clarify the need for this? I believe that this will break twister.

**Comment by smrtos** [comment]
> Can you clarify the need for this? I believe that this will break twister.

Current tests don't use these 2 configs, and by default they are enabled. So output will remain as is.
And in future, twister will be changed accordingly.

(Sorry, I mis-clicked the edit and edited your comment.)

**Comment by nashif** [comment]
> Can you clarify the need for this? I believe that this will break twister.

twister will rely on the summary rather than parse the traditional output once this tis ready. This will solve and issue we have with testing devices where console output is delayed and where twister just get the last part of the test output.

**Comment by nashif** [comment]
@smrtos 

getting this output:
> [atomic] total: pass = 2, fail = 0, skip = 0, total = 2, pass rate = 100.00%
[atomic] distinct: pass = 2, fail = 0, skip = 0, total = 2, pass rate = 100.00%
 - [atomic.test_atomic]: pass = 1, fail = 0, skip = 0, total_run = 1
 - [atomic.test_threads_access_atomic]: pass = 1, fail = 0, skip = 0, total_run = 1

The suite output is fine, but the individual tests are a bit too verbose, gicen that each test will only have one result, no? So how about just printing the final status per testcase instead.

**Comment by smrtos** [comment]
> The suite output is fine, but the individual tests are a bit too verbose, given that each test will only have one result, no? So how about just printing the final status per testcase instead.

Each test can run multiple times if `CONFIG_ZTEST_SHUFFLE=y`. 

Totally, a test function can run `CONFIG_ZTEST_SHUFFLE_SUITE_REPEAT_COUNT * CONFIG_ZTEST_SHUFFLE_TEST_REPEAT_COUNT` times. 

By default it's 3x3.

Maybe it's possible that each run has a different result. So I'd like to provide a more in-depth summary.

@nashif

**Comment by chen-png** [comment]
could we use a bitarray to handle testcase's status result?
the total bits could be predefined by shuffle repeat config, and for each testcase, it only needs two bits to handle three status(pass, fail, skip), and if we enable the shuffle and this testcase will be executed many times, we could just shift left two bits for each time, then we could get each testcase' status in every repeated execution, like below:
 - [pipe_api.test_pipe_reader_wait]: P F P
 - [pipe_api.test_pipe_user_get_fail]: P S P
 - [pipe_api.test_pipe_write_avail_null]: P P P
 - [pipe_api.test_resource_pool_auto_free]: P S P

**Comment by nashif** [comment]
> Each test can run multiple times if `CONFIG_ZTEST_SHUFFLE=y`.

This is just going to be a mess if we have to take shuffle into consideration in test reporting, the way I see it, a test should be marked failed if it failed just once, it does not really matter if it passed the 2 other times, this will simplify the reporting and provide a consistent output and result.

**Comment by smrtos** [comment]
> > Each test can run multiple times if `CONFIG_ZTEST_SHUFFLE=y`.
> 
> This is just going to be a mess if we have to take shuffle into consideration in test reporting, the way I see it, a test should be marked failed if it failed just once, it does not really matter if it passed the 2 other times, this will simplify the reporting and provide a consistent output and result.

Got it~ Now the summary looks like this:


```
------ TESTSUITE SUMMARY ------
PASS - 100.00% [mutex_api]: pass =  3, fail =  0, skip =  0, total =  3
 - PASS - [mutex_api.test_mutex_recursive]
 - PASS - [mutex_api.test_mutex_reent_lock_no_wait]
 - PASS - [mutex_api.test_mutex_reent_lock_timeout_fail]

FAIL -  75.00% [mutex_api_1cpu]: pass =  3, fail =  1, skip =  0, total =  4
 - PASS - [mutex_api_1cpu.test_mutex_lock_unlock]
 - FAIL - [mutex_api_1cpu.test_mutex_priority_inheritance]
 - PASS - [mutex_api_1cpu.test_mutex_reent_lock_forever]
 - PASS - [mutex_api_1cpu.test_mutex_reent_lock_timeout_pass]

```

**Comment by smrtos** [comment]
For code like below,

https://github.com/zephyrproject-rtos/zephyr/blob/6ede1da43494da3840b2f270de955f6bf11cbbf0/tests/subsys/portability/cmsis_rtos_v2/src/timer.c#L33-L40

If we disable all test output, variable like `Tmr` will be reported as not used. 
So I add an extra `print_nothing` function: 
https://github.com/smrtos/zephyr/blob/e29f5af6e553b17214d625c814781e5b0bc9ce82/subsys/testsuite/include/zephyr/tc_util.h#L124-L127

File: subsys/testsuite/ztest/include/zephyr/ztest_test_new.h
Comment (yperess) [inline subsys/testsuite/ztest/include/zephyr/ztest_test_new.h:150]: nit: please run clang-format

**Review by yperess — COMMENTED** [review]
This looks great, thanks for all the work!

**Review by yperess — DISMISSED** [review]
Approving from my end, I'll let Anas weigh in on how this will be captured by Twister and used.

@smrtos I'll need to re-approve once you run clang-format on your changes, just re-request review when you do.

File: subsys/testsuite/ztest/include/zephyr/ztest_test_new.h
Comment (smrtos) [reply subsys/testsuite/ztest/include/zephyr/ztest_test_new.h:150]: Will do. Thanks for the reminding.

**Review by chen-png — DISMISSED** [review]

File: subsys/testsuite/ztest/include/zephyr/ztest_test_new.h
Comment (smrtos) [reply subsys/testsuite/ztest/include/zephyr/ztest_test_new.h]: I tried `clang-format-12`. It made a few changes. Some doesn't make sense to me. Such as below one.
So I only adopt the change to above line 150.
```
.after = (after_fn), .teardown = (teardown_fn),   
```

**Comment by smrtos** [comment]
Guys, any more comments?  Please let me know when I still remember the logic.

**Review by chen-png — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by stephanosio — COMMENTED** [review]
Two serious failures started appearing after this was merged:

```
INFO    - 1155/1155 qemu_riscv32_smp          tests/subsys/logging/log_api/logging.log_api_deferred_printk_cpp FAILED Timeout (qemu 59.982s)
s====T=E=S=T=S=U=I=T=E= =t=e=s=t=_=l=o=g=_=a=p=i=_=1=c=p=u= =s=u=c=c=e=e=d=e=d=
=
==========
=
=-=-=-=-=-=-= =T=E=S=T=S=U=I=T=E
 
SUMMARRYP A-S-S- ----


100..0 - PASS -- [tte s[tt_elsotg__laopgi_.atpeis]t:_ lpoags_sa r=g _8e,v aflauialt i=o n0],
```

```
INFO    - 1155/1155 qemu_cortex_a53_smp       tests/kernel/sched/schedule_api/kernel.scheduler   FAILED Timeout (qemu 179.977s)
...
===================================================================
E: ELR_ELn: 0x0000000040039a60
E: ESR_ELn: 0x000000008200000f
E:   EC:  0x20 (Instruction Abort from a lower Exception level, that might be using AArch32 or AArch64)
...
```

https://github.com/zephyrproject-rtos/zephyr/runs/7583447039?check_suite_focus=true#step:11:2704
https://github.com/zephyrproject-rtos/zephyr/runs/7583447039?check_suite_focus=true#step:11:2704

We should revert this.

**Pull request merged**
